From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Stained Glass OS <ke7oxh@gmail.com>
Date: Wed, 14 Oct 2026 23:55:27 -0700
Subject: [PATCH] wintypes: IReference for every scalar PropertyValue type
 (3291)

PropertyValue boxes answer IReference<T> for every scalar type they can be
created from (UInt8, Int16, Int32, UInt32, Int64, UInt64, Single, Guid,
DateTime, TimeSpan, Point, Size, Rect); only boolean, double and string did.
C++/WinRT box_value() and Windows.Web.Http header properties (ContentLength,
Expires, ...) need them. One generic face, IReference<T>::get_Value copies
the stored value.

diff --git a/dlls/wintypes/main.c b/dlls/wintypes/main.c
index 12861c9..7c37fb1 100644
--- a/dlls/wintypes/main.c
+++ b/dlls/wintypes/main.c
@@ -369,6 +369,36 @@ static const struct IApiInformationStaticsVtbl api_information_statics_vtbl =
     api_information_statics_IsApiContractPresentByMajorAndMinor
 };
 
+/* IReference<T> of the other scalar types: all share the layout IInspectable + get_Value(T *) */
+struct generic_iref_vtbl
+{
+    HRESULT (STDMETHODCALLTYPE *QueryInterface)(void *iface, REFIID iid, void **out);
+    ULONG (STDMETHODCALLTYPE *AddRef)(void *iface);
+    ULONG (STDMETHODCALLTYPE *Release)(void *iface);
+    HRESULT (STDMETHODCALLTYPE *GetIids)(void *iface, ULONG *count, IID **iids);
+    HRESULT (STDMETHODCALLTYPE *GetRuntimeClassName)(void *iface, HSTRING *name);
+    HRESULT (STDMETHODCALLTYPE *GetTrustLevel)(void *iface, TrustLevel *level);
+    HRESULT (STDMETHODCALLTYPE *get_Value)(void *iface, void *value);
+};
+static const struct generic_iref_vtbl generic_iref_vtbl;
+
+static const struct { const IID *iid; PropertyType type; } generic_iref_types[] =
+{
+    { &IID_IReference_BYTE, PropertyType_UInt8 },
+    { &IID_IReference_INT16, PropertyType_Int16 },
+    { &IID_IReference_INT32, PropertyType_Int32 },
+    { &IID_IReference_UINT32, PropertyType_UInt32 },
+    { &IID_IReference_INT64, PropertyType_Int64 },
+    { &IID_IReference_UINT64, PropertyType_UInt64 },
+    { &IID_IReference_FLOAT, PropertyType_Single },
+    { &IID_IReference_GUID, PropertyType_Guid },
+    { &IID_IReference_DateTime, PropertyType_DateTime },
+    { &IID_IReference_TimeSpan, PropertyType_TimeSpan },
+    { &IID_IReference_Point, PropertyType_Point },
+    { &IID_IReference_Size, PropertyType_Size },
+    { &IID_IReference_Rect, PropertyType_Rect },
+};
+
 struct property_value
 {
     IPropertyValue IPropertyValue_iface;
@@ -377,6 +407,7 @@ struct property_value
         IReference_boolean boolean_iface;
         IReference_DOUBLE double_iface;
         IReference_HSTRING hstring_iface;
+        struct { const struct generic_iref_vtbl *lpVtbl; } generic;
     } irefs;
     PropertyType type;
     unsigned int value_size;
@@ -493,6 +524,22 @@ static HRESULT STDMETHODCALLTYPE property_value_QueryInterface(IPropertyValue *i
         return S_OK;
     }
 
+    else
+    {
+        unsigned int i;
+
+        for (i = 0; i < ARRAY_SIZE(generic_iref_types); i++)
+        {
+#ifdef SG_MUTANT_WINTYPES_IREF
+            continue;
+#endif
+            if (!IsEqualIID(riid, generic_iref_types[i].iid) || impl->type != generic_iref_types[i].type) continue;
+            IPropertyValue_AddRef(iface);
+            *out = &impl->irefs.generic;
+            return S_OK;
+        }
+    }
+
     WARN("%s not implemented, returning E_NOINTERFACE.\n", debugstr_guid(riid));
     *out = NULL;
     return E_NOINTERFACE;
@@ -537,6 +584,7 @@ static HRESULT STDMETHODCALLTYPE property_value_GetIids(IPropertyValue *iface, U
     struct property_value *impl = impl_from_IPropertyValue(iface);
     const IID *list[2] = { &IID_IPropertyValue };
     ULONG count = 1;
+    unsigned int i;
 
     TRACE("iface %p, iid_count %p, iids %p.\n", iface, iid_count, iids);
 
@@ -546,6 +594,9 @@ static HRESULT STDMETHODCALLTYPE property_value_GetIids(IPropertyValue *iface, U
         list[count++] = &IID_IReference_DOUBLE;
     else if (impl->type == PropertyType_String)
         list[count++] = &IID_IReference_HSTRING;
+    else
+        for (i = 0; i < ARRAY_SIZE(generic_iref_types); i++)
+            if (impl->type == generic_iref_types[i].type) list[count++] = generic_iref_types[i].iid;
     return return_iids(iid_count, iids, list, count);
 }
 
@@ -887,6 +938,7 @@ static HRESULT _create_primitive_property_value(PropertyType type, void *value,
         return E_OUTOFMEMORY;
 
     impl->IPropertyValue_iface.lpVtbl = &property_value_vtbl;
+    impl->irefs.generic.lpVtbl = &generic_iref_vtbl;
     impl->type = type;
     impl->ref = 1;
     if (type != PropertyType_Empty)
@@ -1005,6 +1057,55 @@ static const struct IReference_DOUBLEVtbl iref_double_vtbl =
     iref_double_get_Value,
 };
 
+static inline struct property_value *impl_from_generic_iref(void *iface)
+{
+    return CONTAINING_RECORD(iface, struct property_value, irefs.generic);
+}
+static HRESULT STDMETHODCALLTYPE generic_iref_QueryInterface(void *iface, REFIID iid, void **out)
+{
+    return IPropertyValue_QueryInterface(&impl_from_generic_iref(iface)->IPropertyValue_iface, iid, out);
+}
+static ULONG STDMETHODCALLTYPE generic_iref_AddRef(void *iface)
+{
+    return IPropertyValue_AddRef(&impl_from_generic_iref(iface)->IPropertyValue_iface);
+}
+static ULONG STDMETHODCALLTYPE generic_iref_Release(void *iface)
+{
+    return IPropertyValue_Release(&impl_from_generic_iref(iface)->IPropertyValue_iface);
+}
+static HRESULT STDMETHODCALLTYPE generic_iref_GetIids(void *iface, ULONG *count, IID **iids)
+{
+    return IPropertyValue_GetIids(&impl_from_generic_iref(iface)->IPropertyValue_iface, count, iids);
+}
+static HRESULT STDMETHODCALLTYPE generic_iref_GetRuntimeClassName(void *iface, HSTRING *name)
+{
+    return IPropertyValue_GetRuntimeClassName(&impl_from_generic_iref(iface)->IPropertyValue_iface, name);
+}
+static HRESULT STDMETHODCALLTYPE generic_iref_GetTrustLevel(void *iface, TrustLevel *level)
+{
+    return IPropertyValue_GetTrustLevel(&impl_from_generic_iref(iface)->IPropertyValue_iface, level);
+}
+static HRESULT STDMETHODCALLTYPE generic_iref_get_Value(void *iface, void *value)
+{
+    struct property_value *impl = impl_from_generic_iref(iface);
+
+    TRACE("iface %p, value %p.\n", iface, value);
+
+    if (!value) return E_POINTER;
+    memcpy(value, impl->value, impl->unit_size);
+    return S_OK;
+}
+static const struct generic_iref_vtbl generic_iref_vtbl =
+{
+    generic_iref_QueryInterface,
+    generic_iref_AddRef,
+    generic_iref_Release,
+    generic_iref_GetIids,
+    generic_iref_GetRuntimeClassName,
+    generic_iref_GetTrustLevel,
+    generic_iref_get_Value,
+};
+
 DEFINE_IINSPECTABLE(property_value_statics, IPropertyValueStatics, struct wintypes, IActivationFactory_iface)
 
 static HRESULT STDMETHODCALLTYPE property_value_statics_CreateEmpty(IPropertyValueStatics *iface,
